Sources/NIO/CircularBuffer.swift

```diff
@@ -50,36 +50,63 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
     ///
     /// Amortized *O(1)*
     public mutating func append(_ value: E) {
+        return self.addLast(value)
+    }
+
+    /// Add an element to the end of the ring buffer.
+    ///
+    /// Amortized *O(1)*
+    public mutating func addLast(_ value: E) {
         self.buffer[self.tailIdx] = value
         self.tailIdx = (self.tailIdx + 1) & self.mask
 
         if self.headIdx == self.tailIdx {
             // No more room left for another append so grow the buffer now.
-            var newBacking: ContiguousArray<E?> = []
-            let newCapacity = self.buffer.count << 1 // Double the storage.
-            precondition(newCapacity > 0, "Can't double capacity of \(self.buffer.count)")
-            assert(newCapacity % 2 == 0)
-
-            newBacking.reserveCapacity(newCapacity)
-            newBacking.append(contentsOf: self.buffer[self.headIdx..<self.buffer.count])
-            if self.headIdx > 0 {
-                newBacking.append(contentsOf: self.buffer[0..<self.headIdx])
-            }
-            newBacking.append(contentsOf: repeatElement(nil, count: newCapacity - newBacking.count))
-            self.tailIdx = self.buffer.count
-            self.headIdx = 0
-            self.buffer = newBacking
+            self.doubleCapacity()
+        }
+    }
+
+    /// Add an element to the front of the ring buffer.
+    ///
+    /// Amortized *O(1)*
+    public mutating func addFirst(_ value: E) {
+        let idx = (self.headIdx - 1) & mask
+        self.buffer[idx] = value
+        self.headIdx = idx
+
+        if self.headIdx == self.tailIdx {
+            // No more room left for another append so grow the buffer now.
+            self.doubleCapacity()
         }
     }
 
+    /// Double the capacity of the buffer and adjust the headIdx and tailIdx.
+    private mutating func doubleCapacity() {
+        var newBacking: ContiguousArray<E?> = []
+        let newCapacity = self.buffer.count << 1 // Double the storage.
+        precondition(newCapacity > 0, "Can't double capacity of \(self.buffer.count)")
+        assert(newCapacity % 2 == 0)
+
+        newBacking.reserveCapacity(newCapacity)
+        newBacking.append(contentsOf: self.buffer[self.headIdx..<self.buffer.count])
+        if self.headIdx > 0 {
+            newBacking.append(contentsOf: self.buffer[0..<self.headIdx])
+        }
+        newBacking.append(contentsOf: repeatElement(nil, count: newCapacity - newBacking.count))
+        self.tailIdx = self.buffer.count
+        self.headIdx = 0
+        self.buffer = newBacking
+    }
+
     /// Remove the front element of the ring buffer.
     ///
     /// *O(1)*
     public mutating func removeFirst() -> E {
-        precondition(!self.isEmpty)
-
         let value = self.buffer[self.headIdx]
-        self.buffer[headIdx] = nil
+
+        precondition(value != nil, "CircularBuffer is empty")
```

Comment:
this should be the first instruction I think, just check `isEmpty`

Reply:
@weissi sure I can do this but I would argue that this way we it may be a bit faster...

Reply:
If you want a speedup, the better thing to do is to do this:

```swift
guard let value = self.buffer[headIdx] else {
    preconditionFailure("Circular buffer is empty")
}
self.buffer[headIdx] = nil
self.headIdx = (self.headIdx + 1) & self.mask
return value
```

This checks the optional only once, not twice.

Reply:
fine with me as well...(would use fatalError tho) @weissi  ?

Reply:
happy with either